File: library/src/main/java/uk/co/barbuzz/zigzagrecyclerview/ZigzagGridRecyclerViewAdapter.java
package uk.co.barbuzz.zigzagrecyclerview;

import android.content.Context;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;

import com.squareup.picasso.Picasso;

import java.util.ArrayList;
import java.util.List;

public class ZigzagGridRecyclerViewAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private static final int IMAGE_VIEW_EVEN_TYPE = 0;
    private static final int IMAGE_VIEW_ODD_TYPE = 1;

    private List<? extends ZigzagImage> zigzagImageList = new ArrayList<>();
    private Context context;
    private ZigzagListOnClickListener zigzagListOnClickListener;
    private int backgroundColourResId;
    private int placeholderDrawableResId;

    public interface ZigzagListOnClickListener {
        void onZigzagImageClicked(int position, ZigzagImage zigzagImage);
    }

    public ZigzagGridRecyclerViewAdapter(Context context, List<? extends ZigzagImage> zigzagImageList,
                                         ZigzagListOnClickListener zigzagListOnClickListener) {
        this.context = context;
        this.zigzagImageList = zigzagImageList;
        this.zigzagListOnClickListener = zigzagListOnClickListener;
        backgroundColourResId = context.getResources().getColor(R.color.oblique_background);
    }

    @Override
    public int getItemViewType(int position) {
        return position % 2 == 0 ? IMAGE_VIEW_EVEN_TYPE : IMAGE_VIEW_ODD_TYPE;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view;
        if (viewType == IMAGE_VIEW_ODD_TYPE) {
            view = LayoutInflater.from(parent.getContext()).inflate(R.layout.element_oblique_views, parent, false);
        } else {
            view = LayoutInflater.from(parent.getContext()).inflate(R.layout.element_oblique_views_2, parent, false);
        }
        return new ImageViewHolder(view);
    }

    @Override
    public void onBindViewHolder(final RecyclerView.ViewHolder viewHolder, final int position) {
        final ImageViewHolder imageViewHolder = (ImageViewHolder) viewHolder;

        if (backgroundColourResId != 0) {
            imageViewHolder.obliqueViewParentLayout.setBackgroundColor(backgroundColourResId);
        }

        //load first image
        int pos1 = position + position;
        final ZigzagImage zigzagImage1 = zigzagImageList.get(pos1);
        String imageUrl1 = zigzagImage1.getZigzagImageUrl();
        if (imageUrl1 == null || imageUrl1.trim().equalsIgnoreCase("")) {
            int imageResource1 = zigzagImage1.getZigzagImageResourceId();
            loadImageResource(imageResource1, imageViewHolder.obliqueView1);
        } else {
            loadImageResourceUrl(imageUrl1, imageViewHolder.obliqueView1);
        }
        imageViewHolder.obliqueView1.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                zigzagListOnClickListener.onZigzagImageClicked(position, zigzagImage1);
            }
        });

        //load second image if need as we might have odd number in list
        int secondImagePos = pos1 + 1;
        if (secondImagePos < zigzagImageList.size()) {
            final ZigzagImage zigzagImage2 = zigzagImageList.get(secondImagePos);
            if (zigzagImage2 != null) {
                String imageUrl2 = zigzagImage2.getZigzagImageUrl();
                if (imageUrl1 == null || imageUrl1.trim().equalsIgnoreCase("")) {
                    int imageResource2 = zigzagImage2.getZigzagImageResourceId();
                    loadImageResource(imageResource2, imageViewHolder.obliqueView2);
                } else {
                    loadImageResourceUrl(imageUrl2, imageViewHolder.obliqueView2);
                }
                imageViewHolder.obliqueView2.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        zigzagListOnClickListener.onZigzagImageClicked(position, zigzagImage2);
                    }
                });
            }
        } else {
            imageViewHolder.obliqueView2.setVisibility(View.INVISIBLE);
        }
    }

    @Override
    public int getItemCount() {
        float size = zigzagImageList.size();
        int count = Math.round(size / 2);
        return count;
    }

    private void loadImageResource(int imageResource, ImageView obliqueView) {
        if (placeholderDrawableResId==0) {
            Picasso.with(context).load(imageResource)
                    .into(obliqueView);
        } else {
            Picasso.with(context).load(imageResource)
                    .placeholder(placeholderDrawableResId)
                    .into(obliqueView);
        }
    }

    private void loadImageResourceUrl(String imageUrlString, ImageView obliqueView) {
        if (placeholderDrawableResId==0) {
            Picasso.with(context).load(imageUrlString)
                    .into(obliqueView);
        } else {
            Picasso.with(context).load(imageUrlString)
                    .placeholder(placeholderDrawableResId)
                    .into(obliqueView);
        }
    }

    public void setData(List<? extends ZigzagImage> zigzagImageList) {
        this.zigzagImageList = zigzagImageList;
        notifyDataSetChanged();
    }

    public List<? extends ZigzagImage> getData() {
        return zigzagImageList;
    }

    public ZigzagListOnClickListener getZigzagListOnClickListener() {
        return zigzagListOnClickListener;
    }

    public void setZigzagListOnClickListener(ZigzagListOnClickListener zigzagListOnClickListener) {
        this.zigzagListOnClickListener = zigzagListOnClickListener;
    }

    public void setBackgroundColourResId(int backgroundColourResId) {
        this.backgroundColourResId = backgroundColourResId;
    }

    public void setPlaceholderDrawableResId(int placeholderDrawableResId) {
        this.placeholderDrawableResId = placeholderDrawableResId;
    }

    public class ImageViewHolder extends RecyclerView.ViewHolder {
        LinearLayout obliqueViewParentLayout;
        ImageView obliqueView1;
        ImageView obliqueView2;

        public ImageViewHolder(View v) {
            super(v);

            obliqueViewParentLayout = v.findViewById(R.id.obliqueViewLayout);
            obliqueView1 = v.findViewById(R.id.obliqueView);
            obliqueView2 = v.findViewById(R.id.obliqueView2);
        }
    }

}

File: library/src/main/java/uk/co/barbuzz/zigzagrecyclerview/ZigzagGridRecyclerViewAdapter.kt
package uk.co.barbuzz.zigzagrecyclerview

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import androidx.recyclerview.widget.RecyclerView
import com.squareup.picasso.Picasso
import java.util.*

class ZigzagGridRecyclerViewAdapter(context: Context, zigzagImageList: List<ZigzagImage>,
                                    zigzagListOnClickListener: ZigzagListOnClickListener) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    private var zigzagImageList: List<ZigzagImage> = ArrayList()
    var zigzagListOnClickListener: ZigzagListOnClickListener
    private var backgroundColourResId: Int
    private var placeholderDrawableResId = 0

    interface ZigzagListOnClickListener {
        fun onZigzagImageClicked(position: Int, zigzagImage: ZigzagImage?)
    }

    override fun getItemViewType(position: Int): Int {
        return if (position % 2 == 0) IMAGE_VIEW_EVEN_TYPE else IMAGE_VIEW_ODD_TYPE
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val view: View
        view = if (viewType == IMAGE_VIEW_ODD_TYPE) {
            LayoutInflater.from(parent.context).inflate(R.layout.element_oblique_views, parent, false)
        } else {
            LayoutInflater.from(parent.context).inflate(R.layout.element_oblique_views_2, parent, false)
        }
        return ImageViewHolder(view)
    }

    override fun onBindViewHolder(viewHolder: RecyclerView.ViewHolder, position: Int) {
        val imageViewHolder = viewHolder as ImageViewHolder
        if (backgroundColourResId != 0) {
            imageViewHolder.obliqueViewParentLayout.setBackgroundColor(backgroundColourResId)
        }
        //load first image
        val pos1 = position + position
        val zigzagImage1 = zigzagImageList[pos1]
        val imageUrl1 = zigzagImage1?.zigzagImageUrl
        if (imageUrl1 == null || imageUrl1.trim { it <= ' ' }.equals("", ignoreCase = true)) {
            val imageResource1 = zigzagImage1.zigzagImageResourceId
            loadImageResource(imageResource1, imageViewHolder.obliqueView1)
        } else {
            loadImageResourceUrl(imageUrl1, imageViewHolder.obliqueView1)
        }
        imageViewHolder.obliqueView1.setOnClickListener { zigzagListOnClickListener.onZigzagImageClicked(position, zigzagImage1) }
        //load second image if need as we might have odd number in list
        val secondImagePos = pos1 + 1
        if (secondImagePos < zigzagImageList.size) {
            val zigzagImage2 = zigzagImageList[secondImagePos]
            if (zigzagImage2 != null) {
                val imageUrl2 = zigzagImage2.zigzagImageUrl
                if (imageUrl1 == null || imageUrl1.trim { it <= ' ' }.equals("", ignoreCase = true)) {
                    val imageResource2 = zigzagImage2.zigzagImageResourceId
                    loadImageResource(imageResource2, imageViewHolder.obliqueView2)
                } else {
                    loadImageResourceUrl(imageUrl2, imageViewHolder.obliqueView2)
                }
                imageViewHolder.obliqueView2.setOnClickListener { zigzagListOnClickListener.onZigzagImageClicked(position, zigzagImage2) }
            }
        } else {
            imageViewHolder.obliqueView2.visibility = View.INVISIBLE
        }
    }

    override fun getItemCount(): Int {
        val size = zigzagImageList.size.toFloat()
        return Math.round(size / 2)
    }

    private fun loadImageResource(imageResource: Int, obliqueView: ImageView) {
        if (placeholderDrawableResId == 0) {
            Picasso.get().load(imageResource)
                    .into(obliqueView)
        } else {
            Picasso.get().load(imageResource)
                    .placeholder(placeholderDrawableResId)
                    .into(obliqueView)
        }
    }

    private fun loadImageResourceUrl(imageUrlString: String?, obliqueView: ImageView) {
        if (placeholderDrawableResId == 0) {
            Picasso.get().load(imageUrlString)
                    .into(obliqueView)
        } else {
            Picasso.get().load(imageUrlString)
                    .placeholder(placeholderDrawableResId)
                    .into(obliqueView)
        }
    }

    var data: List<ZigzagImage>
        get() = zigzagImageList
        set(zigzagImageList) {
            this.zigzagImageList = zigzagImageList
            notifyDataSetChanged()
        }

    fun setBackgroundColourResId(backgroundColourResId: Int) {
        this.backgroundColourResId = backgroundColourResId
    }

    fun setPlaceholderDrawableResId(placeholderDrawableResId: Int) {
        this.placeholderDrawableResId = placeholderDrawableResId
    }

    inner class ImageViewHolder(v: View) : RecyclerView.ViewHolder(v) {
        var obliqueViewParentLayout: LinearLayout
        var obliqueView1: ImageView
        var obliqueView2: ImageView

        init {
            obliqueViewParentLayout = v.findViewById(R.id.obliqueViewLayout)
            obliqueView1 = v.findViewById(R.id.obliqueView)
            obliqueView2 = v.findViewById(R.id.obliqueView2)
        }
    }

    companion object {
        private const val IMAGE_VIEW_EVEN_TYPE = 0
        private const val IMAGE_VIEW_ODD_TYPE = 1
    }

    init {
        this.zigzagImageList = zigzagImageList
        this.zigzagListOnClickListener = zigzagListOnClickListener
        backgroundColourResId = context.resources.getColor(R.color.oblique_background)
    }
}